### PR TITLE
Add Consumes endpoint constraint

### PR DESCRIPTION
**Pull request by JamesNK**
* Add Consumes endpoint constraint
* Add copy of existing Consumes action constraint unit tests for endpoint constraint
* Run Consumes functional tests with both routing and dispatching

Fixes https://github.com/aspnet/Routing/issues/586

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs [hunk 1]
```diff
@@ -184,6 +192,83 @@ public bool Accept(ActionConstraintContext context)
...
+            // In case there is a single candidate with a constraint it should be selected.
+            // If there are multiple actions with consumes action constraints this should result in ambiguous exception
+            // unless there is another action without a consumes constraint.
+            if (requestContentType == null)
+            {
+                var isActionWithoutConsumeConstraintPresent = context.Candidates.Any(
+                    candidate => candidate.Constraints == null ||
+                    !candidate.Constraints.Any(constraint => constraint is IConsumesEndpointConstraint));
+
+                return !isActionWithoutConsumeConstraintPresent;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs]: Is most of this just copy-paste?

This kind of logic is something we'll have to think about - though I'm not sure if we did it anywhere else than consumes.

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs]: For instance we might handle this in MVC by creating a 'sentinel' endpoint for this case that always returns a 415.

Or we might build a first class feature into dispatcher for this 🤔

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ConsumesAttributeTests.cs [hunk 2]
```diff
@@ -293,6 +295,193 @@ public void Accept_ForNoRequestType_ReturnsTrueForAllConstraints(string contentT
...
+            };
+
+            context.HttpContext = CreateHttpContext(contentType: contentType);
+
+            // Act & Assert
+            context.CurrentCandidate = context.Candidates[0];
+            Assert.True(constraint1.Accept(context));
+            context.CurrentCandidate = context.Candidates[1];
+            Assert.True(constraint2.Accept(context));
+        }
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ConsumesAttributeTests.cs]: I'm assuming these are clones of the existing tests and not looking at them in depth

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/ConsumesAttributeTestsBase.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+using BasicWebSite.Models;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Testing.xunit;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.FunctionalTests
+{
+    public abstract class ConsumesAttributeTestsBase<TStartup> : IClassFixture<MvcTestFixture<TStartup>> where TStartup : class
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ConsumesAttributeTestsBase.cs]: nice

File: src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs (hunk 1 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs]: It's copy-paste.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ConsumesAttributeTests.cs (hunk 2 above)
Comment (JamesNK) [reply test/Microsoft.AspNetCore.Mvc.Core.Test/ConsumesAttributeTests.cs]: +1 copy-paste.

File: src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs (hunk 1 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Mvc.Core/ConsumesAttribute.cs]: I'm going to change `action` -> `endpoint`. Are there other changes you want to make here in the future? Do we want an issue?

**Pull request merged**